kernel/kernel-api/src/main/java/io/delta/kernel/internal/compaction/LogCompactionWriter.java

```diff
@@ -0,0 +1,162 @@
+  private static final Logger logger = LoggerFactory.getLogger(LogCompactionWriter.class);
+
+  private final Path logPath;
+  private final long startVersion;
+  private final long endVersion;
+  private final long minFileRetentionTimestampMillis;
+
+  public LogCompactionWriter(
+      Path logPath, long startVersion, long endVersion, long minFileRetentionTimestampMillis) {
+    this.logPath = requireNonNull(logPath);
+    this.startVersion = startVersion;
+    this.endVersion = endVersion;
+    this.minFileRetentionTimestampMillis = minFileRetentionTimestampMillis;
+  }
+
+  public void writeLogCompactionFile(Engine engine) throws IOException {
+    String fileName = String.format("%020d.%020d.compacted.json", startVersion, endVersion);
+    Path compactedPath = new Path(logPath, fileName);
+
+    logger.info(
+        "Writing compacted log file for versions {} to {} to path: {}",
+        startVersion,
+        endVersion,
+        compactedPath);
+
+    final long startTimeMillis = System.currentTimeMillis();
+    final List<FileStatus> deltas =
+        DeltaLogActionUtils.listDeltaLogFilesAsIter(
+                engine,
+                new HashSet<>(Arrays.asList(DeltaLogFileType.COMMIT)),
+                logPath,
+                startVersion,
+                Optional.of(endVersion),
+                true /* mustBeRecreatable */)
+            .toInMemoryList();
+
+    logger.info(
+        "{}: Took {}ms to list commit files for log compaction",
+        logPath,
+        System.currentTimeMillis() - startTimeMillis);
+
+    if (deltas.isEmpty()) {
+      logger.warn(
+          "Asked to do a log compaction between {} and {}, " + "but there are no files to compact",
+          startVersion,
+          endVersion);
+      return;
+    }
+
+    final long lastCommitTimestamp = getLast(deltas).getModificationTime();
+
+    LogSegment segment =
+        new LogSegment(logPath, endVersion, deltas, emptyList(), lastCommitTimestamp);
+    CreateCheckpointIterator checkpointIterator =
+        new CreateCheckpointIterator(engine, segment, minFileRetentionTimestampMillis);
+    wrapEngineExceptionThrowsIO(
+        () -> {
+          try (CloseableIterator<Row> rows = new FilteredBatchToRowIter(checkpointIterator)) {
+            engine.getJsonHandler().writeJsonFileAtomically(compactedPath.toString(), rows, false);
+          }
+          logger.info("Successfully wrote compacted log file `{}`", compactedPath);
+          return null;
+        },
+        "Wrote compacted log file `%s`",
+        compactedPath);
+  }
+
+  /** Utility to determine if log compaction should run for the given commit version. */
+  public static boolean shouldCompact(long commitVersion, long compactionInterval) {
+    return commitVersion > 0 && (commitVersion % compactionInterval == 0);
+  }
+
+  /** Utility to convert an Iterator<FilteredColumnarBatch> into an Iterator<Row> */
+  private static class FilteredBatchToRowIter implements CloseableIterator<Row> {
```

Comment:
I think this could just be distilled to a `flatMap` right? @vkorukanti do you know if we ever added this utility somewhere I think we've discussed it a few times

Reply:
It totally can! But we don't  have flatMap available afaict? Java has it for `Stream` but our iterators don't implement that.

Reply:
We would implement our own version of it on CloseableIterator (our own interface anyways) or in a utilities class but I thought maybe we had done this? If not maybe we can add this as a util here? I know we've wanted this before

Reply:
Okay. I'm checking with @vkorukanti if we already have this, otherwise I can add it.

Reply:
We don't have one. Planning to add one.

However in my opinion we should change or add our JsonHandler to also take the ColumnarBatch and create json rows out of the batches. Delta sharing found that instead of getting rows if they used column vectors in batch directly, it was significant (~40% better) perf improvement.

Reply:
Thanks Venki!

Sounds like this is still a discussion point. @allisonport-db thoughts on just having this utility here and then moving it to a more general spot if/when we settle on exactly what we want to do?

Reply:
Sounds good, move to `Utils`